backend/app/main.py

```diff
@@ -139,15 +137,58 @@ async def update_assignments(
     stmt = stmt.on_conflict_do_update(
         constraint=Assignments.__table__.primary_key, set_={"zone": stmt.excluded.zone}
     )
-    session.exec(stmt)
+    session.execute(stmt)
     session.commit()
     return {"assignments_upserted": len(data.assignments)}
 
 
+@app.patch(
+    "/api/update_assignments/{document_id}/shatter_parents",
+    response_model=ShatterResult,
+)
+async def shatter_parent(
+    document_id: str, data: GEOIDS, session: Session = Depends(get_session)
+):
+    stmt = text("""SELECT *
+        FROM shatter_parent(:input_document_id, :parent_geoids)""").bindparams(
+        bindparam(key="input_document_id", type_=UUIDType),
+        bindparam(key="parent_geoids", type_=ARRAY(String)),
+    )
+    results = session.execute(
+        statement=stmt,
+        params={
+            "input_document_id": document_id,
+            "parent_geoids": data.geoids,
+        },
+    )
+    # :( was getting validation errors so am just going to loop
+    assignments = [
+        Assignments(document_id=str(document_id), geo_id=geo_id, zone=zone)
+        for document_id, geo_id, zone in results
+    ]
+    result = ShatterResult(parents=data, children=assignments)
+    session.commit()
+    return result
+
+
 # called by getAssignments in apiHandlers.ts
-@app.get("/api/get_assignments/{document_id}", response_model=list[Assignments])
+@app.get("/api/get_assignments/{document_id}", response_model=list[AssignmentsResponse])
 async def get_assignments(document_id: str, session: Session = Depends(get_session)):
-    stmt = select(Assignments).where(Assignments.document_id == document_id)
+    stmt = (
+        select(
+            Assignments.geo_id,
+            Assignments.zone,
+            Assignments.document_id,
+            ParentChildEdges.parent_path,
+        )
+        .join(
+            ParentChildEdges,
+            Assignments.geo_id == ParentChildEdges.child_path,
+            isouter=True,
+        )
+        .where(Assignments.document_id == document_id)
```

Comment:
@nofurtherinformation this is close but once we have multiple sets of parentchildedges (one set per shatterable districtr map), we'll get a larger return set. We need to add a filter on the ParentChildEdges with the following logic:
```sql
SELECT
  doc.document_id,
  doc.geo_id,
  doc.zone,
  edges.parent_path
FROM "document.assignments_<uuid>" doc
LEFT JOIN parentchildedges edges
ON edges.geo_id = child_path
WHERE doc.document_id = :document_id
  AND edges.districtr_map = ( SELECT gerrydb_table FROM document.document WHERE document_id = :document_id )
```

Reply:
Ah yep, good catch. I forgot that these are not per-gerrydb tables / views

Reply:
This actually has a little more to it -- we need to get the DistrictrMap UUID not just the gerrydb table to filter parent child edges, so we need to snag districtrmap  (`districtrmap.gerrydb_table_name = document.gerrydb_table`) after getting the document (`document.document_id = :document_id`). 

Something like:

```sql
SELECT 
 assignments.geo_id, 
 assignments.zone, 
 assignments.document_id, 
 edges.parent_path 
FROM assignments 
LEFT OUTER JOIN edges 
  ON assignments.geo_id = edges.child_path 
JOIN districtrmap 
  ON edges.districtr_map = districtrmap.uuid 
JOIN document 
  ON assignments.document_id = document.document_id 
WHERE 
  assignments.document_id = :document_id
AND 
  districtrmap.gerrydb_table_name = document.gerrydb_table
```

Could also be subqueries, or a UDF if we wanted

Reply:
Looks about right – slightly updated the query. Right now I've enforced that the gerrydb_table_name is unique per districtrmap, which we probably don't want in the long term, but it simplifies the query. The query above (and implemented with the ORM) was failing a tests in BE when there was no parentchildedges